File: apps/sim-bridge/Sources/SimBridge/Protocol.swift
import Foundation

// `DeviceInfo` and `DeviceState` declare `Codable` at their definition site in
// `Simulator/ServiceContext.swift` — Swift blocks automatic synthesis from a
// non-declaring file.

// MARK: - Incoming (pane → daemon)

public enum PaneToBridge: Sendable {
    case deviceList
    case deviceBoot(udid: String)
    case deviceShutdown(udid: String)
    case deviceInstall(udid: String, appPath: String)
    case deviceLaunch(udid: String, bundleId: String)
    case inputTap(x: Double, y: Double, phase: TapPhase)
    case inputDrag(points: [DragPoint])
    case inputKey(usage: Int32, down: Bool, modifiers: Int32)
    case inputButton(kind: HardwareButton, down: Bool)
    case axEnable
    case axHit(x: Double, y: Double, mode: AXHitMode)
    case axTree
    case axSnapshot
    case axAction(elementId: String, action: String)
    case rotate(orientation: Int)
    case unknown

    public enum TapPhase: String, Codable, Sendable { case down, up }
    public enum AXHitMode: String, Codable, Sendable { case hover, select }

    public struct DragPoint: Codable, Sendable {
        public let x: Double
        public let y: Double
        public let t: Double
    }
}

extension PaneToBridge: Decodable {
    private enum CodingKeys: String, CodingKey {
        case type, udid, appPath, bundleId, x, y, phase, points, usage, down, modifiers, kind, elementId, action, orientation, mode
    }
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let t = try c.decode(String.self, forKey: .type)
        switch t {
        case "deviceList": self = .deviceList
        case "deviceBoot": self = .deviceBoot(udid: try c.decode(String.self, forKey: .udid))
        case "deviceShutdown": self = .deviceShutdown(udid: try c.decode(String.self, forKey: .udid))
        case "deviceInstall":
            self = .deviceInstall(udid: try c.decode(String.self, forKey: .udid),
                                   appPath: try c.decode(String.self, forKey: .appPath))
        case "deviceLaunch":
            self = .deviceLaunch(udid: try c.decode(String.self, forKey: .udid),
                                 bundleId: try c.decode(String.self, forKey: .bundleId))
        case "inputTap":
            self = .inputTap(
                x: try c.decode(Double.self, forKey: .x),
                y: try c.decode(Double.self, forKey: .y),
                phase: try c.decode(TapPhase.self, forKey: .phase)
            )
        case "inputDrag":
            self = .inputDrag(points: try c.decode([DragPoint].self, forKey: .points))
        case "inputKey":
            self = .inputKey(
                usage: try c.decode(Int32.self, forKey: .usage),
                down: try c.decode(Bool.self, forKey: .down),
                modifiers: try c.decodeIfPresent(Int32.self, forKey: .modifiers) ?? 0
            )
        case "inputButton":
            let kindRaw = try c.decode(String.self, forKey: .kind)
            let kind = HardwareButton(rawValue: kindRaw) ?? .home
            self = .inputButton(kind: kind, down: try c.decode(Bool.self, forKey: .down))
        case "axEnable": self = .axEnable
        case "axHit":
            self = .axHit(
                x: try c.decode(Double.self, forKey: .x),
                y: try c.decode(Double.self, forKey: .y),
                mode: try c.decodeIfPresent(AXHitMode.self, forKey: .mode) ?? .select
            )
        case "axTree": self = .axTree
        case "axSnapshot": self = .axSnapshot
        case "axAction":
            self = .axAction(
                elementId: try c.decode(String.self, forKey: .elementId),
                action: try c.decode(String.self, forKey: .action)
            )
        case "rotate":
            self = .rotate(orientation: try c.decode(Int.self, forKey: .orientation))
        default:
            self = .unknown
        }
    }
}

// MARK: - Outgoing (daemon → pane)

public enum BridgeToPane: Sendable {
    case deviceListResponse(devices: [DeviceInfo])
    case deviceState(udid: String, state: DeviceState, bootStatus: String?)
    case displayReady(contextId: UInt32, pixelWidth: Int, pixelHeight: Int, scale: Double)
    case displaySurfaceChanged(pixelWidth: Int, pixelHeight: Int)
    case axHitResponse(chain: [AXElement], hitIndex: Int, mode: PaneToBridge.AXHitMode)
    case axTreeResponse(root: AXElement)
    case axSnapshotResponse(nodes: [AXNode], appContext: SimAppInfo?)
    case error(code: String, message: String, detail: [String: String]?)
}

extension BridgeToPane: Encodable {
    private enum CodingKeys: String, CodingKey {
        case type, devices, udid, state, bootStatus, contextId, pixelWidth, pixelHeight, scale, chain, hitIndex, root, code, message, detail, mode, nodes, appContext
    }
    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .deviceListResponse(devices):
            try c.encode("deviceListResponse", forKey: .type)
            try c.encode(devices, forKey: .devices)
        case let .deviceState(udid, state, bootStatus):
            try c.encode("deviceState", forKey: .type)
            try c.encode(udid, forKey: .udid)
            try c.encode(state, forKey: .state)
            try c.encodeIfPresent(bootStatus, forKey: .bootStatus)
        case let .displayReady(contextId, w, h, scale):
            try c.encode("displayReady", forKey: .type)
            try c.encode(contextId, forKey: .contextId)
            try c.encode(w, forKey: .pixelWidth)
            try c.encode(h, forKey: .pixelHeight)
            try c.encode(scale, forKey: .scale)
        case let .displaySurfaceChanged(w, h):
            try c.encode("displaySurfaceChanged", forKey: .type)
            try c.encode(w, forKey: .pixelWidth)
            try c.encode(h, forKey: .pixelHeight)
        case let .axHitResponse(chain, hitIndex, mode):
            try c.encode("axHitResponse", forKey: .type)
            try c.encode(chain, forKey: .chain)
            try c.encode(hitIndex, forKey: .hitIndex)
            try c.encode(mode, forKey: .mode)
        case let .axTreeResponse(root):
            try c.encode("axTreeResponse", forKey: .type)
            try c.encode(root, forKey: .root)
        case let .axSnapshotResponse(nodes, appContext):
            try c.encode("axSnapshotResponse", forKey: .type)
            try c.encode(nodes, forKey: .nodes)
            try c.encodeIfPresent(appContext, forKey: .appContext)
        case let .error(code, message, detail):
            try c.encode("error", forKey: .type)
            try c.encode(code, forKey: .code)
            try c.encode(message, forKey: .message)
            try c.encodeIfPresent(detail, forKey: .detail)
        }
    }
}

File: packages/sim-pane/src/protocol.ts
export const SIM_BRIDGE_DEFAULT_PORT = 17323;

export type DeviceState =
  | "shutdown"
  | "booting"
  | "booted"
  | "shuttingDown"
  | "creating"
  | "unknown";

export interface DeviceInfo {
  udid: string;
  name: string;
  runtime: string;
  model: string;
  state: DeviceState;
}

export interface AXFrame {
  x: number;
  y: number;
  width: number;
  height: number;
  /** Corner radius of the backing iOS view (display points). Forwarded to
   *  the native picker so the outline shape matches the element — rounded
   *  for squircle buttons and material cards, zero for plain rectangles. */
  cornerRadius?: number;
}

export interface SimAppInfo {
  bundleId: string;
  name: string | null;
  pid: number;
  bundlePath: string | null;
  dataContainer: string | null;
  executablePath: string | null;
  projectPath: string | null;
}

export interface AXSourceHint {
  absolutePath: string;
  line: number;
  reason: string;
  confidence: number;
  /** Raw source neighborhood (newline-joined) centered on `line`. Only set
   *  when the bridge could read the file; unresolved hints stay terse. */
  snippet?: string | null;
  /** 1-indexed file line matching the first row of `snippet`. */
  snippetStartLine?: number | null;
}

export interface AXElement {
  id: string;
  role: string;
  label: string | null;
  value: string | null;
  frame: AXFrame;
  identifier: string | null;
  enabled: boolean;
  selected: boolean;
  children: AXElement[] | null;
  appContext: SimAppInfo | null;
  sourceHints: AXSourceHint[] | null;
}

/** Flat BFS node for the full-snapshot feed. Mirrors `AXNode` in the bridge
 *  — the pane rebuilds topology via `parentId` instead of shipping nested
 *  trees, which keeps the outline overlay + gap math cheap. */
export interface AXNode {
  id: string;
  parentId: string | null;
  role: string;
  label: string | null;
  value: string | null;
  identifier: string | null;
  frame: AXFrame;
  enabled: boolean;
  selected: boolean;
}

/** Parsed from `.inspectable()`'s `accessibilityIdentifier` stamp.
 *  Format: `ModuleName/File.swift:42` or `ModuleName/File.swift:42|name=Alias`. */
export interface InspectableAnchor {
  module: string | null;
  file: string;
  line: number;
  alias: string | null;
  sourcePath: string;
  absolutePath: string | null;
}

export type AXHitMode = "hover" | "select";

export type HardwareButtonKind = "home" | "lock" | "siri" | "side" | "applePay" | "keyboard";

export type PaneToBridge =
  | { type: "deviceList" }
  | { type: "deviceBoot"; udid: string }
  | { type: "deviceShutdown"; udid: string }
  | { type: "inputTap"; x: number; y: number; phase: "down" | "up" }
  | { type: "inputDrag"; points: { x: number; y: number; t: number }[] }
  | { type: "inputKey"; usage: number; down: boolean; modifiers: number }
  | { type: "inputButton"; kind: HardwareButtonKind; down: boolean }
  | { type: "axEnable" }
  | { type: "axHit"; x: number; y: number; mode: AXHitMode }
  | { type: "axTree" }
  | { type: "axSnapshot" }
  | { type: "axAction"; elementId: string; action: string }
  | { type: "rotate"; orientation: number };

export type BridgeToPane =
  | { type: "deviceListResponse"; devices: DeviceInfo[] }
  | { type: "deviceState"; udid: string; state: DeviceState; bootStatus: string | null }
  | {
      type: "displayReady";
      contextId: number;
      pixelWidth: number;
      pixelHeight: number;
      scale: number;
    }
  | { type: "displaySurfaceChanged"; pixelWidth: number; pixelHeight: number }
  | { type: "axHitResponse"; chain: AXElement[]; hitIndex: number; mode: AXHitMode }
  | { type: "axTreeResponse"; root: AXElement }
  | {
      type: "axSnapshotResponse";
      nodes: AXNode[];
      appContext: SimAppInfo | null;
    }
  | { type: "error"; code: string; message: string; detail?: Record<string, string> };
